src/App.tsx

```diff
@@ -1,24 +1,339 @@
+    } catch (error) {
+      setTodoDeleteError(true);
+    } finally {
+      setTargetTodosIds([]);
+    }
+  };
+
+  const visibleTodos = useMemo(() => {
+    if (!todos) {
+      return null;
+    }
+
+    switch (filteredBy) {
+      case FilteredBy.All:
+        return todos;
+
+      case FilteredBy.Completed:
+        return todos.filter(todo => todo.completed);
+
+      case FilteredBy.Active:
+        return todos.filter(todo => !todo.completed);
+
+      default:
+        return null;
+    }
+  }, [filteredBy, todos]);
+
+  const setFilter = (value: FilteredBy) => {
+    setFilteredBy(value);
+  };
+
+  const handleInput = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setNewTodoTitle(event.currentTarget.value);
+  };
+
+  const changeTodoStatus = async (todoId: number, completed: boolean) => {
+    setTargetTodosIds(currIds => {
+      return [...currIds, todoId || 0];
+    });
+    setTodoUpdatingError(false);
+    try {
+      await patchTodos(todoId, { completed });
+
+      setTodos((currTodo) => {
+        if (currTodo) {
+          return currTodo.map(todo => {
+            if (todo.id === todoId) {
+              return { ...todo, completed };
+            }
+
+            return todo;
+          });
+        }
+
+        return null;
+      });
+    } catch (error) {
+      setTodoUpdatingError(true);
+    } finally {
+      setTargetTodosIds([]);
+    }
+  };
+
+  const toggleAll = async () => {
+    const todosForUpdaiting: Promise<unknown>[] = [];
+    const targetTodoIds: number[] = [];
+
+    if (isActive) {
+      todos?.forEach(todo => {
+        if (!todo.completed) {
+          todosForUpdaiting.push(patchTodos(todo.id, { completed: true }));
+          targetTodoIds.push(todo.id);
+        }
+      });
+    } else {
+      todos?.forEach(todo => {
+        todosForUpdaiting.push(patchTodos(todo.id, { completed: false }));
+        targetTodoIds.push(todo.id);
+      });
+    }
+
+    setTodoUpdatingError(false);
+
+    try {
+      setTargetTodosIds(currIds => currIds.concat(targetTodoIds));
+      await Promise.all(todosForUpdaiting);
+
+      if (isActive) {
+        setTodos(todos?.map(todo => {
+          if (!todo.completed) {
+            return { ...todo, completed: true };
+          }
+
+          return todo;
+        }) || null);
+      } else {
+        setTodos(todos?.map(todo => ({ ...todo, completed: false })) || null);
+      }
+    } catch (error) {
+      setTodoUpdatingError(true);
+    } finally {
+      setTargetTodosIds([]);
+    }
+  };
+
+  const callEditeMode = (title: string) => {
+    setUpdatingTitle(title);
+  };
+
+  const submitTitleUpdating = async (todoId: number, oldTitle: string) => {
+    setTodoUpdatingError(false);
+
+    if (oldTitle === updatingTitle) {
+      return;
+    }
+
+    setTargetTodosIds(currIds => {
+      return [...currIds, todoId || 0];
+    });
+    if (updatingTitle !== '') {
+      try {
+        setTodos((currTodo) => {
```

Comment:
You first update the state, and then make a request, and this is wrong, todo should't update if the request passed with an error